src/eduid_userdb/group_management/db.py

```diff
@@ -0,0 +1,174 @@
+#        copyright notice, this list of conditions and the following
+#        disclaimer in the documentation and/or other materials provided
+#        with the distribution.
+#     3. Neither the name of the NORDUnet nor the names of its
+#        contributors may be used to endorse or promote products derived
+#        from this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+#
+import datetime
+import logging
+from dataclasses import replace
+from operator import itemgetter
+from typing import Any, Dict, List, Optional
+
+from eduid_userdb.db import BaseDB
+from eduid_userdb.exceptions import DocumentOutOfSync
+from eduid_userdb.group_management.state import GroupInviteState
+
+logger = logging.getLogger(__name__)
+
+
+class GroupManagementInviteStateDB(BaseDB):
+    def __init__(self, db_uri: str, db_name: str = 'eduid_group_management', collection: str = 'group_invite_data'):
+        super(GroupManagementInviteStateDB, self).__init__(db_uri, db_name, collection=collection)
+        # Create an index so that invites for group_scim_id, email_address and role is unique
+        indexes = {
+            'unique-group-email-role': {
+                'key': [('group_scim_id', 1), ('email_address', 1), ('role', 1)],
+                'unique': True,
+            }
+        }
+        self.setup_indexes(indexes)
+
+    def get_state(
+        self, group_scim_id: str, email_address: str, role: str, raise_on_missing: bool = True
+    ) -> Optional[GroupInviteState]:
+        """
+        :param group_scim_id: Groups unique identifier
+        :param email_address: Invited email address
+        :param role: Group role
+        :param raise_on_missing: Raise exception if True else return None
+
+        :return: GroupInviteState instance
+        """
+        spec = {'group_scim_id': group_scim_id, 'email_address': email_address, 'role': role}
+        docs = list(self._get_documents_by_filter(spec, raise_on_missing=raise_on_missing))
+
+        if len(docs) > 1:
+            # Ex. multiple states for same group_scim_id, email address and roles matched
+            # This should not be possible but we have seen it happen
```

Comment:
comment copied from elsewhere, where there was no index?

given that there is an index, maybe just abort if this happens?